### DISCUSSION TITLE
Solidify a surface voxel

### DISCUSSION THREAD

**Discussion by SimonNgj** [AcademySoftwareFoundation/openvdb#1401] [General]
I'm having a point cloud. After converting it to a FloatGrid, I have a surface voxel model. I want to obtain a solid voxel model. Is there any way I can do that with OpenVDB? I'm thinking of Floodfill 3D algorithm with a requirement that the surface voxel model needs to be watertight. Any suggestion is appreciated

**Comment by Idclip** [comment]
There are signed flood fill methods available in `openvdb/tools/SignedFloodFill.h`

**Comment by SimonNgj** [comment]
I tried that but I didn't receive what I expected. The VDBs before and after applying SignedFloodFill are exactly similar, still only surface voxels. I don't know what SignedFloodFill actually does here. I wonder does it really FloodFill algorithm: https://en.wikipedia.org/wiki/Flood_fill
![Capture](https://user-images.githubusercontent.com/22513384/170272896-2811c982-904e-492b-ba4c-90327cb6451b.PNG)

**Comment by Idclip** [comment]
I think I see what you're after now - the signed flood fill agorithms fill inside/outside voxels with _constant_ values and do not modify active states. They are indeed a flood fill algorithm, but it sounds like what you want to do is activate the interior of your level set with exact distances? These algorithms do not re-compute the distance values.

There are a variety of ways to re-construct the interior of a level set, I think the easiest is perhaps just to use `openvdb::tools::levelSetRebuild()`. This will convert your level set to polygons and reconstruct the interior up to the specified distance. If you want to completely fill the interior with distances you can simply use a large value:

```
ls = levelSetRebuild(*ls, 0, /*ext=*/halfWidth, /*interior=*/std::numeric_limits<float>::max()); // LevelSetRebuild.h
```

Alternatively there are tools in `FastSweeping.h` which can be used to re-compute interior values and avoid level set reconstruction. @kmuseth / @apradhana can probably elaborate on those.

**Comment by SimonNgj** [comment]
Thank @Idclip, yes, I want to activate the interior bound by the surface voxel. I tried 
`ls = levelSetRebuild(*ls, 0, /*ext=*/halfWidth, /*interior=*/std::numeric_limits<float>::max()); // LevelSetRebuild.h`
and saw the increase in the number of active voxels, but when I visualize it by vdb_view.exe, I didn't see the difference.

**Comment by apradhana** [comment]
@SimonNgj: I tried the problem you're describing by scattering points on the Stanford's bunny model.
<img width="591" alt="Screen Shot 2022-05-25 at 10 57 44 PM" src="https://user-images.githubusercontent.com/4682534/170426797-840e627d-5dde-4734-b0ac-4aeb0ec85de2.png">

In the example that I showed, I use this [dilate](https://github.com/AcademySoftwareFoundation/openvdb/blob/master/openvdb/openvdb/tools/FastSweeping.h#L414) function

**Comment by SimonNgj** [comment]
Hi @apradhana, I followed your instruction, but end up with an empty VDB. Could you please help me to check what is wrong here:
![image](https://user-images.githubusercontent.com/22513384/170612731-245e0b17-9f90-4ba4-bfec-42518bbbdff8.png)

**Comment by SimonNgj** [comment]
It seems like signedFloodFill function is to estimate the distance value, not to fill the watertight surface

**Comment by apradhana** [comment]
@SimonNgj Would you mind clarifying what you mean by "filling a watertight surface"? What voxel values are you looking for the interior voxels?

(One possible answer is that you want the interior voxels to have a uniform value of, say, 1.f. If that is so, you can use the topology you get from `levelSetRebuild` to rewrite the values of the interior voxels.)

**Comment by SimonNgj** [comment]
Hi @apradhana, I want to achieve a solid voxelization like this: https://orbilu.uni.lu/bitstream/10993/16275/1/GARCIA_Frederic_ICPR2014.pdf
After converting a point cloud to voxels, we get a surface voxel representation of the point cloud.

**Comment by SimonNgj** [comment]
This image shows what I want to achieve: starting from a point cloud (particles) => surface voxelization (level set) => solid voxelization
![image](https://user-images.githubusercontent.com/22513384/171330968-e4f0eb7c-3d71-4407-9032-9cb0f497e7fa.png)

**Comment by SimonNgj** [comment]
Hi @apradhana, does a higher interior bandwidth take a longer time to execute?

**Reply by apradhana** [reply]
More half-band width means more active voxels. So it will take a longer time to execute.
